### PR TITLE
Add filtering mechanism for the sensitive data (#3207)

### PR DESCRIPTION
**Pull request by sbbroot**

### CODE REVIEW

**Review by seriva — DISMISSED** [review]

File: ansible/playbooks/roles/repository/library/filter_credentials.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        required: true
+        type: str
+    dest:
+        description: Path to the newly created, filtered manifest
+        required: false
+        type: str
+    in_place:
+        description: When set to True will modify existing manifest passed as `src`
+        required: false
+        type: false
```
Comment (to-bar) [inline ansible/playbooks/roles/repository/library/filter_credentials.py]:
```suggestion
        type: bool
        default: false
```

**Comment by seriva** [comment]
@to-bar I think this filter should cover everything, I went quickly through the defaults and besides **cloud** and **...password...** I don't think anything more is there. Or maybe you know something more:)

@sbbroot Maybe we can also apply this to the dump_debug_info function in epicli.py:

image
image

Would be nice to kill 2 birds with one stone.

File: ansible/playbooks/roles/repository/library/filter_credentials.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+import yaml
+from ansible.module_utils.basic import AnsibleModule
+
+
+def _get_filtered_manifest(manifest_path: Path) -> str:
+    """
+    Load the manifest file and remove any sensitive data.
+
+    :param manifest_path: manifest file which will be loaded
+    :returns: filtered manifset doc
```
Comment (to-bar) [inline ansible/playbooks/roles/repository/library/filter_credentials.py]:
```suggestion
    :returns: filtered manifest
```

File: ansible/playbooks/roles/repository/library/filter_credentials.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+        for line in mhandler.readlines():
+            if 'password' not in line.lower():
+                lines.append(line)
+
+    docs = yaml.safe_load_all(''.join(lines))
+    epiphany_cluster = next(docs)
+
+    # remove provider credentials:
+    try:
+        del epiphany_cluster['specification']['cloud']
```
Comment (to-bar) [inline ansible/playbooks/roles/repository/library/filter_credentials.py]: How about removing only `credentials` subkey?

File: ansible/playbooks/roles/repository/tasks/copy-download-requirements.yml [hunk 4]
```diff
@@ -55,9 +55,16 @@
         dest: "{{ download_requirements_dir }}/{{ item }}"
       loop: "{{ _files }}"
 
+    - name: Filter sensitive data from the manifest
+      filter_credentials:
+        src: "{{ input_manifest_path }}"
+        dest: /tmp/filtered_manifest.yml
+      delegate_to: localhost
+      when: not full_download and input_manifest_path
```
Comment (to-bar) [inline ansible/playbooks/roles/repository/tasks/copy-download-requirements.yml:59]: The next task has the same condition so I would suggest to use `block` to have just one `when`.

**Review by to-bar — CHANGES_REQUESTED** [review]

**Comment by to-bar** [comment]
> @to-bar I think this filter should cover everything, I went quickly through the defaults and besides **cloud** and **...password...** I don't think anything more is there. Or maybe you know something more:)

I have `grep`-ed `schema` subdir using `secret` pattern. Found some keys but by default they are commented out, e.g:
```
  - name: rabbitmq
    #image_pull_secret_name: regcred # optional
```
```
#  image_registry_secrets:
#  - email: emaul@domain.com
#    name: secretname
#    namespace: default
#    password: docker-registry-pwd
#    server_url: docker-registry-url
#    username: docker-registry-user
```
Maybe emails should also be filtered out?

```
#     - name: 'opsgenie'
#       opsgenie_config:
#         api_key: <secret> | default = global.opsgenie_api_key
#         api_url: <string> | default = global.opsgenie_api_url
```

File: ansible/playbooks/roles/repository/library/filter_credentials.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
+#!/usr/bin/python
+
+from __future__ import (absolute_import, division, print_function)
+
+from pathlib import Path
+from typing import Callable, Dict, List
```
Comment (to-bar) [inline ansible/playbooks/roles/repository/library/filter_credentials.py]: We support only Python 3.10 so you can consider using just `list` and `dict`.

File: ansible/playbooks/roles/repository/library/filter_credentials.py [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+                                                            'configuration/feature-mappings',
+                                                            'configuration/image-registry']]
+
+    FILTER_DATA: Dict[str, Callable] = {
+        'any': _filter_common,
+        'azure': _filter_azure,
+        'aws': _filter_aws
+    }
+
+    FILTER_DATA[filtered_docs[0]['provider']](filtered_docs)
```
Comment (to-bar) [inline ansible/playbooks/roles/repository/library/filter_credentials.py:105]: @seriva can input manifest contain multiple providers or we always have just one?

File: ansible/playbooks/roles/repository/library/filter_credentials.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        module.fail_json(msg='Cannot use `in_place` and `dest` at the same time', **result)
+
+    if not module.params['in_place'] and not module.params['dest']:  # to stdout
+        result['manifest'] = manifest
+    elif module.params['in_place']:  # overwrite existing manifest
+        with input_manifest.open(mode='w', encoding='utf-8') as mhandler:
+            mhandler.write(manifest)
+    elif module.params['dest']:  # write to a new location
+        with Path(module.params['dest']).open(mode='w', encoding='utf-8') as output_manifest:
+            output_manifest.write(manifest)
```
Comment (to-bar) [inline ansible/playbooks/roles/repository/library/filter_credentials.py]:
```suggestion
        with Path(module.params['dest']).open(mode='w', encoding='utf-8') as output_file:
            output_file.write(manifest)
```

File: ansible/playbooks/roles/repository/library/filter_credentials.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+    result['changed'] = True
```
Comment (to-bar) [inline ansible/playbooks/roles/repository/library/filter_credentials.py]: IMO to support this properly we should compare content before and after filtering. If manifest was already filtered the task should not be marked as changed as content would be the same.

File: .devcontainer/python.env [hunk 9]
```diff
@@ -2,4 +2,4 @@
 
 # PYTHONPATH can contain multiple locations separated by os.pathsep: semicolon (;) on Windows and colon (:) on Linux/macOS.
 # Invalid paths are ignored. To verify use "python.analysis.logLevel": "Trace".
-PYTHONPATH=ansible/playbooks/roles/repository/files/download-requirements
+PYTHONPATH=ansible/playbooks/roles/repository:ansible/playbooks/roles/repository/files/download-requirements
```
Comment (to-bar) [inline .devcontainer/python.env]:
```suggestion
PYTHONPATH=ansible/playbooks/roles/repository/library:ansible/playbooks/roles/repository/files/download-requirements
```

File: ansible/playbooks/roles/repository/library/filter_credentials.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+    :param manifest_path: manifest file which will be loaded
+    :returns: filtered manifset
+    """
+    docs = yaml.safe_load_all(manifest_path.open())
+    filtered_docs = [doc for doc in docs if doc['kind'] in ['epiphany-cluster',
+                                                            'configuration/feature-mappings',
+                                                            'configuration/image-registry']]
+
+    FILTER_DATA: list[str, Callable] = {
```
Comment (to-bar) [inline ansible/playbooks/roles/repository/library/filter_credentials.py]:
```suggestion
    FILTER_DATA: dict[str, Callable] = {
```

**Review by to-bar — CHANGES_REQUESTED** [review]

File: ansible/playbooks/roles/repository/library/filter_credentials.py (hunk 6 above)
Comment (sbbroot) [reply ansible/playbooks/roles/repository/library/filter_credentials.py:105]: We support only single provider in a manifest.

**Review by to-bar — DISMISSED** [review]

**Review by to-bar — DISMISSED** [review]

**Comment by seriva** [comment]
> @to-bar I think this filter should cover everything, I went quickly through the defaults and besides **cloud** and **...password...** I don't think anything more is there. Or maybe you know something more:)
> 
> @sbbroot Maybe we can also apply this to the dump_debug_info function in epicli.py:
> 
> image image
> 
> Would be nice to kill 2 birds with one stone.

As discussed with @sbbroot lets defer this to a separate issue as this is not entirely related to this issue. @sbbroot pls create it:)

**Review by seriva — DISMISSED** [review]

**Review by przemyslavic — APPROVED** [review]

**Review by seriva — APPROVED** [review]

**Pull request merged**